pkg/controllers/resources/pods/validatePodSecurity.go

```diff
@@ -0,0 +1,156 @@
+	admissionv1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/pod-security-admission/admission"
+	admissionapi "k8s.io/pod-security-admission/admission/api"
+	"k8s.io/pod-security-admission/api"
+	"k8s.io/pod-security-admission/metrics"
+	"k8s.io/pod-security-admission/policy"
+)
+
+func (s *podSyncer) isPodSecurityStandardsValid(ctx *synccontext.SyncContext, pod *corev1.Pod) (bool, error) {
+	result, err := s.validatePodSecurityStandards(ctx.Context, pod)
+	fmt.Println(result)
+	if err != nil {
+		ctx.Log.Errorf(err.Error())
+	} else if result != nil {
+		if !result.Allowed {
+			ctx.Log.Errorf("%s pod creation not allowed: %s", pod.Name, result.Result.Message)
+			pod.Status.Message = result.Result.Message
+			pod.Status.Reason = string(result.Result.Reason)
+			_ = ctx.VirtualClient.Status().Update(ctx.Context, pod)
+			s.EventRecorder().Eventf(pod, "Warning", "SyncError", `Pod %s is forbidden: violates PodSecurity "%s:latest"`, pod.Name, s.podSecurityStandard)
+		}
+		return result.Allowed, nil
+	}
+	return false, err
+}
+
+func (s *podSyncer) validatePodSecurityStandards(ctx context.Context, pod *corev1.Pod) (*admissionv1.AdmissionResponse, error) {
+	if _, ok := podsecurity.PodSecurityExemptNamespaces[pod.Namespace]; ok {
+		return &admissionv1.AdmissionResponse{
+			Allowed: true,
+		}, nil
+	}
+
+	version := api.LatestVersion()
+	podSecurityDefaults := admissionapi.PodSecurityDefaults{
+		Enforce:        s.podSecurityStandard,
+		EnforceVersion: version.String(),
+		Audit:          s.podSecurityStandard,
+		AuditVersion:   version.String(),
+		Warn:           s.podSecurityStandard,
+		WarnVersion:    version.String(),
+	}
+
+	namespaceLister := &namespaceLister{
+		client: s.virtualClusterClient,
+		ctx:    ctx,
+	}
+
+	evaluator, err := policy.NewEvaluator(policy.DefaultChecks())
+	if err != nil {
+		return nil, fmt.Errorf("could not create PodSecurityRegistry: %w", err)
+	}
+
+	adm := &admission.Admission{
+		Configuration: &admissionapi.PodSecurityConfiguration{
+			Defaults: podSecurityDefaults,
+		},
+		Evaluator:        evaluator,
+		Metrics:          &FakeMetricsRecorder{},
+		PodSpecExtractor: admission.DefaultPodSpecExtractor{},
+		PodLister:        admission.PodListerFromClient(s.virtualClusterClient),
+		NamespaceGetter:  admission.NamespaceGetterFromListerAndClient(namespaceLister, s.virtualClusterClient),
+	}
+
+	if err := adm.CompleteConfiguration(); err != nil {
+		return nil, fmt.Errorf("configuration error: %w", err)
+	}
+
+	if err := adm.ValidateConfiguration(); err != nil {
+		return nil, fmt.Errorf("invalid configuration: %w", err)
+	}
+
+	nsPolicy, _ := adm.PolicyToEvaluate(map[string]string{})
```

Comment:
can we safely ignore the error here?

Reply:
I think yes. Anyway, we're using an empty map there.